src/storage/txn/scheduler.rs

```diff
@@ -11,92 +11,593 @@
+    }
+}
+
+fn make_write_cb(pr: ProcessResult, cid: u64, ch: SchedCh) -> Callback<()> {
+    Box::new(move |result: EngineResult<()>| {
+        if let Err(e) = ch.send(Msg::WriteFinish{
+            cid: cid,
+            pr: pr,
+            result: result,
+        }) {
+            error!("write engine failed cmd id {}, err {:?}",
+            cid,
+            e);
+        }
+    })
+}
 
 pub struct Scheduler {
-    store: Arc<TxnStore>,
-    pool: ThreadPool,
+    engine: Arc<Box<Engine>>,
+
+    // cid -> context
+    cmd_ctxs: HashMap<u64, RunningCtx>,
+
+    //
+    schedch: SchedCh,
+
+    // cmd id generator
+    idalloc: u64,
+
+    // simulate memory row locks
+    rowlocks: MemRowLocks,
+
+    // statistics for flow control
+    running_cmd_count: u64,
+}
+
+fn readonly_cmd(cmd: &Command) -> bool {
+    match *cmd {
+        Command::Get { .. } |
+        Command::BatchGet { .. } |
+        Command::Scan { .. } => true,
+        _ => false,
+    }
 }
 
 impl Scheduler {
-    pub fn new(engine: Arc<Box<Engine>>) -> Scheduler {
+    pub fn new(engine: Arc<Box<Engine>>, schedch: SchedCh, concurrency: usize) -> Scheduler {
         Scheduler {
-            store: Arc::new(TxnStore::new(engine)),
-            pool: ThreadPool::new_with_name(thd_name!("txn-scheduler-pool"), DEFAULT_POOL_SIZE),
+            engine: engine,
+            cmd_ctxs: HashMap::new(),
+            schedch: schedch,
+            idalloc: 0,
+            rowlocks: MemRowLocks::new(concurrency),
+            running_cmd_count: 0,
         }
     }
 
-    pub fn exec(&self, cmd: Command) {
-        let store = self.store.clone();
-        self.pool.execute(move || handle_cmd(store, cmd));
+    fn next_id(&mut self) -> u64 {
+        self.idalloc += 1;
+        self.idalloc
     }
-}
 
-fn handle_cmd(store: Arc<TxnStore>, cmd: Command) {
-    let cmd_str = format!("{}", cmd);
-    debug!("scheduler::handle_cmd begin: {}", cmd_str);
-    match cmd {
-        Command::Get { ctx, key, start_ts, callback } => {
-            callback(store.get(ctx, &key, start_ts).map_err(::storage::Error::from));
-        }
-        Command::BatchGet { ctx, keys, start_ts, callback } => {
-            callback(match store.batch_get(ctx, &keys, start_ts) {
-                Ok(results) => {
-                    let mut res = vec![];
-                    for (k, v) in keys.into_iter().zip(results.into_iter()) {
-                        match v {
-                            Ok(Some(x)) => res.push(Ok((k.raw().unwrap(), x))),
-                            Ok(None) => {}
-                            Err(e) => res.push(Err(::storage::Error::from(e))),
-                        }
+    fn incr_running_cmd(&mut self) {
+        self.running_cmd_count += 1;
+    }
+
+    fn decr_running_cmd(&mut self) {
+        self.running_cmd_count -= 1;
+    }
+
+    fn save_cmd_context(&mut self, cid: u64, ctx: RunningCtx) {
+        if let Some(_) = self.cmd_ctxs.insert(cid, ctx) {
```

Comment:
we can return an Option, if Option is not none, we will call the callback outer. 
So we can remove Option<Callback>